### PR TITLE
[BACKLOG-15858] Review recent documentation

### PR DESCRIPTION
**Pull request by dcleao**
@pentaho/millenniumfalcon please review.

### CODE REVIEW

File: impl/client/src/main/doc-js/pentaho/lang/IEventSource.jsdoc [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+ *
+ * To remove an event registration,
+ * it is sufficient to call the [dispose]{@link pentaho.lang.IEventRegistrationHandle#dispose} method
+ * of the registration handle returned by [on]{@link pentaho.lang.IEventSource#on},
+ * upon registration.
+ * Alternatively, as a convenience syntax,
+ * the registration handle can be passed as the single argument to this method.
+ *
+ * It is safe to unregister from an event type while it is being emitted.
+ * However, removed registrations are still taken into account in the current emission.
```
Comment (carlosrusso) [inline impl/client/src/main/doc-js/pentaho/lang/IEventSource.jsdoc]: Perhaps
> However, any registrations removed during the current emission will still execute.

File: impl/client/src/main/javascript/web/pentaho/type/_type.js [hunk 2]
```diff
@@ -901,9 +887,9 @@ define([
       /**
        * Gets the style classes of this and any base types.
        *
-       * Do **NOT** modify the returned array.
+       * Do **not** modify the returned array.
        *
-       * @type string[]
+       * @type {string[]}
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/type/_type.js]: Not that it matters much, but did we settle on a preferred choice between `string[]` and `Array.<string>` ?

File: impl/client/src/main/doc-js/pentaho/type/action/ITarget.jsdoc [hunk 3]
```diff
@@ -17,7 +17,10 @@
 /**
  * The `ITarget` interface represents the entity _where_ an
  * [action]{@link pentaho.type.action.Base} takes place,
- * or the pacient/object of action.
+ * or the patient/object of action.
```
Comment (nantunes) [inline impl/client/src/main/doc-js/pentaho/type/action/ITarget.jsdoc]: Not sure the patient metaphor makes it more clear.

Comment (dcleao) [inline impl/client/src/main/doc-js/pentaho/type/action/ITarget.jsdoc]: The _agent_ acts on the _patient_. It's the most correct, direct way to say — not quite metaphor. "object", on the other hand is more tied to the grammatical subject and object classification. It happens that depending on the verb, the agent and patient may be the subject and the object or the other way round.

All that said, I do realize that the "patient" word, in this context, is not very well known.
I would preferably remove "object".

File: impl/client/src/main/javascript/web/pentaho/type/_type.js (hunk 2 above)
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/type/_type.js]: Whatever we write, Js Doc outputs using the generic syntax. I think that the non-generic syntax cannot represent nested types, like this one: `Map.<string, Object>[]`, which can only be written as `Array.<Map.<string, Object>>`. Even so, I'd prefer not being strict with this, and allow both, as the non-generic syntax is easier, shorter and more rooted.

File: impl/client/src/main/javascript/web/pentaho/lang/EventSource.js [hunk 4]
```diff
@@ -533,17 +533,40 @@ define([
...
   }
 
-  function emitQueue(event, queue, type, phase, isCanceled, keyArgs) {
+  /**
+   * Emits an event.
+   *
+   * @this pentaho.lang.EventSource
+   *
+   * @param {object} event - The event payload object.
+   * @param {!Array.<pentaho.lang.IEventObserverRegistration>} queue - The queue of event observer regitrations.
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/lang/EventSource.js]: typo: regitrations -> registrations

File: impl/client/src/main/javascript/web/pentaho/lang/EventSource.js [hunk 5]
```diff
@@ -533,17 +533,40 @@ define([
...
+   * @param {nonEmptyString} type - The event type.
+   * @param {nonEmptyString} phase - The event phase.
+   * @param {function(object):boolean} [isCanceled] - A predicate that indicates if a given event object
+   * is in a canceled state.
+   * @param {function(any, object, nonEmptyString, nonEmptyString)} [errorHandler] -
+   * When specified with a `null` value,
+   * no error handling is performed and errors thrown by listeners are thrown back to this method's caller.
+   * When unspecified or specified as `undefined`, defaults to a function that simply log the listener errors,
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/lang/EventSource.js]: that simply ~log~ logs the

File: impl/client/src/main/javascript/web/pentaho/lang/EventSource.js [hunk 6]
```diff
@@ -533,17 +533,40 @@ define([
...
+   * and let execution continue to following listeners.
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/lang/EventSource.js]: and ~let~ lets execution to **the** following

File: impl/client/src/main/javascript/web/pentaho/lang/EventSource.js [hunk 7]
```diff
@@ -533,17 +533,40 @@ define([
...
+   * The function arguments are: the error, the event, the event type and the event phase.
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/lang/EventSource.js]: If we need to document the callback, then we should create a doclet just for that...

File: impl/client/src/main/javascript/web/pentaho/lang/EventSource.js [hunk 8]
```diff
@@ -569,21 +591,44 @@ define([
...
   }
 
-  function errorHandlerLog(ex, event, type, phase) {
-    logger.log("Event listener of '" + type + ":" + phase + "' failed. \nStack: " + ex.stack);
+  // TODO: check log level!
+
+  /**
+   * Logs an error thrown by an event listener.
+   *
+   * @param {any} error - The thrown value.
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/lang/EventSource.js]: Is the data type really `{any}` ? Or is it a descendant of `Error` ?

File: impl/client/src/main/javascript/web/pentaho/type/ContainerMixin.js [hunk 9]
```diff
@@ -203,9 +203,11 @@ define([
      *
      * @name pentaho.type.ContainerMixin#_createChangeset
      *
+     * @method
+     *
      * @param {!pentaho.type.changes.Transaction} transaction - The owning transaction.
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/type/ContainerMixin.js]: what about: The transaction that owns this changeset.

File: impl/client/src/main/javascript/web/pentaho/type/ContainerMixin.js [hunk 10]
```diff
@@ -203,9 +203,11 @@ define([
...
      *
-     * @return {pentaho.type.changes.Changeset} A changeset of appropriate type.
+     * @return {!pentaho.type.changes.Changeset} A changeset of appropriate type.
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/type/ContainerMixin.js]: A changeset of **the** appropriate type.

File: impl/client/src/main/javascript/web/pentaho/type/action/base.js [hunk 11]
```diff
@@ -475,6 +652,8 @@ define([
        *
        * @param {!pentaho.type.action.ITarget} target - The action's target.
        * @param {pentaho.type.action.IObserver} [executor] - An action observer to act as action controller/executor.
+       * Unlike with normal event/action observers, the functions of executor observers are called with the executor
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/type/action/base.js]: What about: Unlike ~with~ normal

File: impl/client/src/main/javascript/web/pentaho/type/action/base.js [hunk 12]
```diff
@@ -498,6 +677,8 @@ define([
...
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/type/action/base.js]: Idem, what about: Unlike ~with~ normal

File: impl/client/src/main/javascript/web/pentaho/type/action/base.js [hunk 13]
```diff
@@ -764,23 +1004,74 @@ define([
...
 
-      __doPhaseInit: function() {
+      // region Private __executePhase* Methods
+      /**
+       * Executes the **init** phase.
+       *
+       * Changes the state to [init]{@link pentaho.type.action.States.init}
+       * and delegates to [_onPhaseInit]{@link pentaho.type.action.Base#_onPhaseInit}.
+       *
+       * Used by both the synchronous and the asynchronous action kinds.
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/type/action/base.js]: "kinds" seems superfluous. What about: "Used by both synchronous and asynchronous actions."

File: impl/client/src/main/javascript/web/pentaho/type/action/base.js [hunk 14]
```diff
@@ -764,23 +1004,74 @@ define([
...
       },
 
-      __doPhaseWill: function() {
+      /**
+       * Executes the **will** phase.
+       *
+       * Changes the state to [will]{@link pentaho.type.action.States.will}
+       * and delegates to [_onPhaseWill]{@link pentaho.type.action.Base#_onPhaseWill}.
+       *
+       * Used by both the synchronous and the asynchronous action kinds.
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/type/action/base.js]: idem

File: impl/client/src/main/javascript/web/pentaho/type/action/base.js [hunk 15]
```diff
@@ -764,23 +1004,74 @@ define([
...
-      __doPhaseDo: function() {
+      /**
+       * Executes the **do** phase.
+       *
+       * Changes the state to [do]{@link pentaho.type.action.States.do}
+       * and delegates to [_onPhaseDo]{@link pentaho.type.action.Base#_onPhaseDo},
+       * after which, in case the action is still executing,
+       * calls the [_doDefault]{@link pentaho.type.action.Base#_doDefault} method.
+       *
+       * Used by both the synchronous and the asynchronous action kinds.
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/type/action/base.js]: idem

File: impl/client/src/main/javascript/web/pentaho/type/action/base.js [hunk 16]
```diff
@@ -799,7 +1090,24 @@ define([
...
+       * with an `undefined` result.
+       *
+       * Delegates to [_onPhaseFinally]{@link pentaho.type.action.Base#_onPhaseFinally}.
+       * Any error thrown by it is simply caught and logged.
+       *
+       * Finally, the [executor]{@link pentaho.type.action.Base#_executor} object, if any, is released
+       * and the action [promise]{@link pentaho.type.action.Base#promise}, if previously requested,
+       * is resolved.
+       *
+       * Used by both the synchronous and the asynchronous action kinds.
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/type/action/base.js]: idem.

File: impl/client/src/main/javascript/web/pentaho/type/number.js [hunk 17]
```diff
@@ -35,10 +35,10 @@ define([
      *
      * @description Creates a number instance.
      */
-    return Simple.extend(/** @lends pentaho.type.Number# */{
+    var PenNumber = Simple.extend(/** @lends pentaho.type.Number# */{
       /**
        * Gets the underlying number primitive of the value.
-       * @name pentaho.type.Number#value
+       * @name pentaho.type.PenNumber#value
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/type/number.js]: `pentaho.type.PenNumber` ?

File: impl/client/src/main/javascript/web/pentaho/visual/action/select.js [hunk 18]
```diff
@@ -119,45 +164,104 @@ define([
...
+     *
+     * @param {string|pentaho.visual.action.SelectionMode} value - The selection mode designator.
+     * @param {string} argName - The argument name, for error purposes.
+     *
+     * @return {pentaho.type.Function} A selection mode function wrapped in a simple function value.
+     *
+     * @throws {pentaho.lang.ArgumentInvalidError} When `value` is a `string` which is not one of the
+     * [standard selection mode names]{@link pentaho.visual.action.SelectionModes}.
+     *
+     * @throws {pentaho.lang.ArgumentInvalidTypeError} When `value` is not a `string` or a `function`.
```
Comment (carlosrusso) [inline impl/client/src/main/javascript/web/pentaho/visual/action/select.js]: When `value` is ~not~ neither a `string` ~or~ nor  a `function`

File: impl/client/src/main/javascript/web/pentaho/lang/EventSource.js (hunk 7 above)
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/lang/EventSource.js]: You caught me being lazy... Damn 😀

File: impl/client/src/main/javascript/web/pentaho/lang/EventSource.js (hunk 8 above)
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/lang/EventSource.js]: Unfortunately, JavaScript allows throwing any type of value and we don't control what listener functions decide to throw...

File: impl/client/src/main/javascript/web/pentaho/type/number.js (hunk 17 above)
Comment (dcleao) [inline impl/client/src/main/javascript/web/pentaho/type/number.js]: Side-effects of a rename variable refactoring...

**Comment by dcleao** [comment]
Latest reviews were incorporated.

**Pull request merged**